### PR TITLE
Cache previous results of NavigateTo so we can filter quickly as the user types.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 1]
```diff
@@ -60,13 +75,77 @@ internal abstract partial class AbstractNavigateToSearchService
...
+            CancellationToken cancellationToken)
+        {
+            // Searching an entire project.  See if we already performed that same
+            // search with a substring of the current pattern.  if so, we can use
+            // the previous result and just filter that down.  This is useful for
+            // the common case where a user types some pattern, then keeps adding
+            // to it.
+            ImmutableArray<SearchResult> searchResults;
+            if (s_lastProjectSearchCache.TryGetValue(project, out var previousResult) &&
+                pattern.StartsWith(previousResult.Item1))
```
Comment (Therzok) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: Missing StringComparison kind.

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs]: Discussed offline.  Ok with keeping things like this.

**Comment by CyrusNajmabadi** [comment]
@olegtk Question about NavigateTo presenter.  Right now as the user types, the list is always *cleared* and then the new results come in.  This makes things feel a bit slower than desirable.  It would be great if, as you typed, *if* we were just filtering the existing list, if we could just somehow remove items from the existing list, and keep items around that still apply without the intermediary clearing step.

**Comment by CyrusNajmabadi** [comment]
@olegtk This is the VS4Mac experience:

https://www.screencast.com/t/aH95JbcLg

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

**Pull request merged**